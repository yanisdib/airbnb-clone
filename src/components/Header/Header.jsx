import { useState, useEffect } from 'react';
import styled from 'styled-components';

import { Language } from '@styled-icons/material';

import useAutocompletedCities from '../../hooks/useAutocompletedCities';
import useFormField from '../../hooks/useFormField';

import { Container, MenuButton, MenuDropdown, SearchBar } from '..';

import whiteLogo from '../../assets/svg/logo-original-w.svg';
import originalLogo from '../../assets/svg/logo-original.svg';


const StyledHeader = styled.header`
    display: flex;
    position: ${({ isWindowScrolled }) => isWindowScrolled ? 'fixed' : 'absolute'};
    flex-direction: row;
    --header-brand-color: #FF385C;
    background-color: ${({ isWindowScrolled }) => isWindowScrolled ? '#FFFFFF' : 'var(--ui-brand-color)'};
    color: ${({ isWindowScrolled }) => isWindowScrolled ? 'var(--ui-brand-color)' : '#FFFFFF'};
    fill: ${({ isWindowScrolled }) => isWindowScrolled ? 'var(--ui-brand-color)' : '#FFFFFF'};
    padding: 0;
    margin: 0 auto;
    left: 0;
    right: 0;
    transition: all 300ms ease-in-out;
`;

const Wrapper = styled.div`
    display: flex;
    justify-content: center;
    align-items: flex-start;
    width: 100%;
    background-color: transparent;
`;

const Content = styled.div`
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    align-items: flex-start;
    width: 100%;
`;

const Logo = styled.div`
    display: flex;
    flex: 1 0 140px;
`;

const SearchForm = styled.form`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    max-width: 850px;
    margin: 0 auto;
`;

const SearchTabsGroup = styled.div`
    display: flex;
    align-items: center;
    height: 80px;
`;

const SearchTab = styled.div`
    display: flex;
    position: relative;
    text-align: center;
    font-size: 15.5px;
    line-height: 20px;
    padding: 10px 16px;

    &::after {
        content: "";
        position: absolute;
        width: ${({ isSearchTabActive }) => isSearchTabActive ? 17 : 0}px;
        height: 1px;
        bottom: 0;
        left: 0;
        right: 0;
        margin: 0 auto;
        border-bottom: 2px solid #FFFFFF;
    }

    ${({ isSearchTabActive }) => !isSearchTabActive &&
        `&:hover {                
            cursor: pointer;

            &::after {
                content: "";
                position: absolute;
                width: 5px;
                height: 1px;
                bottom: 0;
                left: 0;
                right: 0;
                margin: 0 auto;
                border-bottom: 2px solid white;
                transition: width 300ms ease;
            }
        }`
    }
`;

const UserNav = styled.div`
    display: flex;
    position: relative;
    justify-content: flex-end;
    align-items: center;
    height: 80px;

    &:last-of-type div{
        margin-left: 8px;
    }

    button {
        width: fit-content;
        height: fit-content;
        background: transparent;
        color: inherit;
        padding: 12px;
        border-radius: 10rem;
        font-size: 14px;
        line-height: 18px;
        cursor: pointer;

        &:hover {
            background: rgba(255, 255, 255, 0.2);
            transition: background 10ms linear;
        }

        svg {
            stroke-width: 1px;
            height: 20px;
        }
    }
`;


function Header() {
    const [searchedLocation, setSearchedLocation] = useFormField('');

    const [isMenuDropdownDisplayed, setIsMenuDropdownDisplayed] = useState(false);
    const [isCheckInDateDropdownDisplayed, setIsCheckInDateDropdownDisplayed] = useState(false);
    const [isCheckOutDateDropdownDisplayed, setIsCheckOutDateDropdownDisplayed] = useState(false);
    const [isGuestsDropdownDisplayed, setIsGuestsDropdownDisplayed] = useState(false);
    const [isWindowScrolled, setIsWindowScrolled] = useState(false);
    const [logo, setLogo] = useState(whiteLogo);

    const autocompletedCities = useAutocompletedCities(searchedLocation);

    const changeHeaderOnScroll = () => {
        if (window.scrollY > 5) {
            setIsWindowScrolled(true)
            setTimeout(() => setLogo(originalLogo), 100)
        }
        else {
            setIsWindowScrolled(false)
            setTimeout(() => setLogo(whiteLogo), 100)
        }
    }

    useEffect(() => window.addEventListener('scroll', changeHeaderOnScroll), [])

    return (
        <StyledHeader isWindowScrolled={isWindowScrolled}>
            <Wrapper>
                <Container>
                    <Content>
                        <Logo>
                            <iframe src={logo} width={102} height={80}>
                                <img src={logo} width={102} height={80} />
                            </iframe>
                        </Logo>
                        <SearchForm>
                            {
                                // TODO: onClick the current Active Tab changes with state (1 as default) taking the current selected Tab
                            }
                            {
                                !isWindowScrolled && (
                                    <>
                                        <SearchTabsGroup>
                                            <SearchTab isSearchTabActive={true}>Places to stay</SearchTab>
                                            <SearchTab>Experiences</SearchTab>
                                            <SearchTab>Online Experiences</SearchTab>
                                        </SearchTabsGroup>
                                        <SearchBar
                                            setLocation={setSearchedLocation}
                                            displayCheckInDateDropdown={() => setIsCheckInDateDropdownDisplayed(prevState => !prevState)}
                                            displayCheckOutDateDropdown={() => setIsCheckOutDateDropdownDisplayed(prevState => !prevState)}
                                            displayGuestsDropdown={() => setIsGuestsDropdownDisplayed(prevState => !prevState)}
                                        />
                                    </>
                                )
                            }
                        </SearchForm>
                        <UserNav>
                            {isMenuDropdownDisplayed && <MenuDropdown />}
                            <button>Become a Host</button>
                            <button>
                                <Language />
                            </button>
                            <MenuButton
                                onClick={() => setIsMenuDropdownDisplayed(prevIsMenuDropdownDisplayed => !prevIsMenuDropdownDisplayed)}
                            />
                        </UserNav>
                    </Content>
                </Container>
            </Wrapper>
        </StyledHeader>
    );
}


export default Header;