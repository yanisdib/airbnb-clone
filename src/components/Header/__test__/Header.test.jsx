import { mount } from 'enzyme';
import createMockStore from 'redux-mock-store';

import Header from '../Header';
import { Provider } from 'react-redux';


describe('Header component', () => {
    let wrapper;

    const middlewares = [];
    const initialState = { cities: {} };
    const mockStore = createMockStore(middlewares);
    const store = mockStore(initialState);

    beforeEach(() => {
        wrapper = mount(
            <Provider store={store}>
                <Header />
            </Provider>
        )
    })

    it('renders correctly without crashing', () => expect(wrapper).toMatchSnapshot())
})