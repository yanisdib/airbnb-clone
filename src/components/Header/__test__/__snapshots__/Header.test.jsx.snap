// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders correctly without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Header>
    <styled.header
      isWindowScrolled={false}
    >
      <header
        className="sc-ikJyIC juDsRi"
      >
        <styled.div>
          <div
            className="sc-jJoQJp hmNxcQ"
          >
            <styled.div>
              <div
                className="sc-bdvvtL fNgVow"
              >
                <styled.div>
                  <div
                    className="sc-hiCibw jSfIfU"
                  >
                    <styled.div>
                      <div
                        className="sc-gWXbKe dJCqhj"
                      >
                        <iframe
                          height={80}
                          src="logo-original-w.svg"
                          width={102}
                        >
                          <img
                            height={80}
                            src="logo-original-w.svg"
                            width={102}
                          />
                        </iframe>
                      </div>
                    </styled.div>
                    <styled.form>
                      <form
                        className="sc-cCcXHH fGmsdf"
                      >
                        <styled.div>
                          <div
                            className="sc-cidDSM suFON"
                          >
                            <styled.div
                              isSearchTabActive={true}
                            >
                              <div
                                className="sc-jcFjpl aKKQk"
                              >
                                Places to stay
                              </div>
                            </styled.div>
                            <styled.div>
                              <div
                                className="sc-jcFjpl bwapvn"
                              >
                                Experiences
                              </div>
                            </styled.div>
                            <styled.div>
                              <div
                                className="sc-jcFjpl bwapvn"
                              >
                                Online Experiences
                              </div>
                            </styled.div>
                          </div>
                        </styled.div>
                        <SearchBar
                          displayCheckInDateDropdown={[Function]}
                          displayCheckOutDateDropdown={[Function]}
                          displayGuestsDropdown={[Function]}
                          setCheckInDate={[Function]}
                          setCheckOutDate={[Function]}
                          setGuests={[Function]}
                          setLocation={[Function]}
                        >
                          <styled.div>
                            <div
                              className="sc-hKwDye fttIFk"
                            >
                              <styled.div>
                                <div
                                  className="sc-eCImPb gzpvSy"
                                >
                                  <styled.div>
                                    <div
                                      className="sc-jRQBWg fhNyBZ"
                                    >
                                      <styled.div>
                                        <div
                                          className="sc-gKclnd flvkWv"
                                        >
                                          <styled.div>
                                            <div
                                              className="sc-iCfMLu tVjpf"
                                            >
                                              <label>
                                                Location
                                              </label>
                                              <styled.input
                                                onChange={[Function]}
                                                placeholder="Where are you going?"
                                              >
                                                <input
                                                  className="sc-pVTFL"
                                                  onChange={[Function]}
                                                  placeholder="Where are you going?"
                                                />
                                              </styled.input>
                                            </div>
                                          </styled.div>
                                          <styled.div>
                                            <div
                                              className="sc-furwcr hCuxfS"
                                            />
                                          </styled.div>
                                        </div>
                                      </styled.div>
                                      <styled.div>
                                        <div
                                          className="sc-gKclnd flvkWv"
                                        >
                                          <styled.div
                                            id="checkInDate"
                                            onClick={[Function]}
                                            role="button"
                                          >
                                            <div
                                              className="sc-iCfMLu tVjpf"
                                              id="checkInDate"
                                              onClick={[Function]}
                                              role="button"
                                            >
                                              <label>
                                                Check in
                                              </label>
                                              <Styled(styled.div)>
                                                <div
                                                  className="sc-jrQzAO sc-kDTinF dvDGoe"
                                                >
                                                  Add dates
                                                </div>
                                              </Styled(styled.div)>
                                            </div>
                                          </styled.div>
                                          <styled.div>
                                            <div
                                              className="sc-furwcr hCuxfS"
                                            />
                                          </styled.div>
                                          <styled.div
                                            id="checkOutDate"
                                            onClick={[Function]}
                                            role="button"
                                          >
                                            <div
                                              className="sc-iCfMLu tVjpf"
                                              id="checkOutDate"
                                              onClick={[Function]}
                                              role="button"
                                            >
                                              <label>
                                                Check out
                                              </label>
                                              <Styled(styled.div)>
                                                <div
                                                  className="sc-jrQzAO sc-iqseJM dvDGoe"
                                                >
                                                  Add dates
                                                </div>
                                              </Styled(styled.div)>
                                            </div>
                                          </styled.div>
                                          <styled.div>
                                            <div
                                              className="sc-furwcr hCuxfS"
                                            />
                                          </styled.div>
                                        </div>
                                      </styled.div>
                                      <styled.div>
                                        <div
                                          className="sc-gKclnd flvkWv"
                                        >
                                          <styled.div
                                            id="guests"
                                            onClick={[Function]}
                                            role="button"
                                          >
                                            <div
                                              className="sc-iCfMLu tVjpf"
                                              id="guests"
                                              onClick={[Function]}
                                              role="button"
                                            >
                                              <label>
                                                Guests
                                              </label>
                                              <Styled(styled.div)>
                                                <div
                                                  className="sc-jrQzAO sc-crHmcD dvDGoe"
                                                >
                                                  Add guests
                                                </div>
                                              </Styled(styled.div)>
                                            </div>
                                          </styled.div>
                                          <styled.div>
                                            <div
                                              className="sc-egiyK lcFibs"
                                            >
                                              <styled.button>
                                                <button
                                                  className="sc-bqiRlB fjIBIT"
                                                >
                                                  <Search>
                                                    <StyledIconBase
                                                      iconAttrs={
                                                        Object {
                                                          "fill": "currentColor",
                                                          "xmlns": "http://www.w3.org/2000/svg",
                                                        }
                                                      }
                                                      iconVerticalAlign="middle"
                                                      iconViewBox="0 0 24 24"
                                                    >
                                                      <ForwardRef
                                                        className="StyledIconBase-ea9ulj-0 bsimgx"
                                                        iconAttrs={
                                                          Object {
                                                            "fill": "currentColor",
                                                            "xmlns": "http://www.w3.org/2000/svg",
                                                          }
                                                        }
                                                        iconVerticalAlign="middle"
                                                        iconViewBox="0 0 24 24"
                                                      >
                                                        <svg
                                                          aria-hidden="true"
                                                          className="StyledIconBase-ea9ulj-0 bsimgx"
                                                          fill="currentColor"
                                                          focusable="false"
                                                          viewBox="0 0 24 24"
                                                          xmlns="http://www.w3.org/2000/svg"
                                                        >
                                                          <path
                                                            d="M0 0h24v24H0z"
                                                            fill="none"
                                                          />
                                                          <path
                                                            d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                                                          />
                                                        </svg>
                                                      </ForwardRef>
                                                    </StyledIconBase>
                                                  </Search>
                                                </button>
                                              </styled.button>
                                            </div>
                                          </styled.div>
                                        </div>
                                      </styled.div>
                                    </div>
                                  </styled.div>
                                </div>
                              </styled.div>
                            </div>
                          </styled.div>
                        </SearchBar>
                      </form>
                    </styled.form>
                    <styled.div>
                      <div
                        className="sc-caiLqq guOQvO"
                      >
                        <button>
                          Become a Host
                        </button>
                        <button>
                          <Language>
                            <StyledIconBase
                              iconAttrs={
                                Object {
                                  "fill": "currentColor",
                                  "xmlns": "http://www.w3.org/2000/svg",
                                }
                              }
                              iconVerticalAlign="middle"
                              iconViewBox="0 0 24 24"
                            >
                              <ForwardRef
                                className="StyledIconBase-ea9ulj-0 bsimgx"
                                iconAttrs={
                                  Object {
                                    "fill": "currentColor",
                                    "xmlns": "http://www.w3.org/2000/svg",
                                  }
                                }
                                iconVerticalAlign="middle"
                                iconViewBox="0 0 24 24"
                              >
                                <svg
                                  aria-hidden="true"
                                  className="StyledIconBase-ea9ulj-0 bsimgx"
                                  fill="currentColor"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M0 0h24v24H0z"
                                    fill="none"
                                  />
                                  <path
                                    d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm6.93 6h-2.95a15.65 15.65 0 00-1.38-3.56A8.03 8.03 0 0118.92 8zM12 4.04c.83 1.2 1.48 2.53 1.91 3.96h-3.82c.43-1.43 1.08-2.76 1.91-3.96zM4.26 14C4.1 13.36 4 12.69 4 12s.1-1.36.26-2h3.38c-.08.66-.14 1.32-.14 2 0 .68.06 1.34.14 2H4.26zm.82 2h2.95c.32 1.25.78 2.45 1.38 3.56A7.987 7.987 0 015.08 16zm2.95-8H5.08a7.987 7.987 0 014.33-3.56A15.65 15.65 0 008.03 8zM12 19.96c-.83-1.2-1.48-2.53-1.91-3.96h3.82c-.43 1.43-1.08 2.76-1.91 3.96zM14.34 14H9.66c-.09-.66-.16-1.32-.16-2 0-.68.07-1.35.16-2h4.68c.09.65.16 1.32.16 2 0 .68-.07 1.34-.16 2zm.25 5.56c.6-1.11 1.06-2.31 1.38-3.56h2.95a8.03 8.03 0 01-4.33 3.56zM16.36 14c.08-.66.14-1.32.14-2 0-.68-.06-1.34-.14-2h3.38c.16.64.26 1.31.26 2s-.1 1.36-.26 2h-3.38z"
                                  />
                                </svg>
                              </ForwardRef>
                            </StyledIconBase>
                          </Language>
                        </button>
                        <MenuButton
                          onClick={[Function]}
                        >
                          <styled.div>
                            <div
                              className="sc-ksdxgE FANRq"
                            >
                              <styled.button
                                onClick={[Function]}
                              >
                                <button
                                  className="sc-hBUSln jnntuF"
                                  onClick={[Function]}
                                >
                                  <Styled(Menu)>
                                    <Menu
                                      className="sc-bkkeKt kVHIMK"
                                    >
                                      <StyledIconBase
                                        className="sc-bkkeKt kVHIMK"
                                        iconAttrs={
                                          Object {
                                            "fill": "currentColor",
                                            "xmlns": "http://www.w3.org/2000/svg",
                                          }
                                        }
                                        iconVerticalAlign="middle"
                                        iconViewBox="0 0 24 24"
                                      >
                                        <ForwardRef
                                          className="StyledIconBase-ea9ulj-0 bsimgx sc-bkkeKt kVHIMK"
                                          iconAttrs={
                                            Object {
                                              "fill": "currentColor",
                                              "xmlns": "http://www.w3.org/2000/svg",
                                            }
                                          }
                                          iconVerticalAlign="middle"
                                          iconViewBox="0 0 24 24"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            className="StyledIconBase-ea9ulj-0 bsimgx sc-bkkeKt kVHIMK"
                                            fill="currentColor"
                                            focusable="false"
                                            viewBox="0 0 24 24"
                                            xmlns="http://www.w3.org/2000/svg"
                                          >
                                            <path
                                              d="M0 0h24v24H0z"
                                              fill="none"
                                            />
                                            <path
                                              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                                            />
                                          </svg>
                                        </ForwardRef>
                                      </StyledIconBase>
                                    </Menu>
                                  </Styled(Menu)>
                                  <styled.div>
                                    <div
                                      className="sc-fotOHu fOvtEr"
                                    >
                                      <Styled(AccountCircle)>
                                        <AccountCircle
                                          className="sc-fFeiMQ hZAAEL"
                                        >
                                          <StyledIconBase
                                            className="sc-fFeiMQ hZAAEL"
                                            iconAttrs={
                                              Object {
                                                "fill": "currentColor",
                                                "xmlns": "http://www.w3.org/2000/svg",
                                              }
                                            }
                                            iconVerticalAlign="middle"
                                            iconViewBox="0 0 24 24"
                                          >
                                            <ForwardRef
                                              className="StyledIconBase-ea9ulj-0 bsimgx sc-fFeiMQ hZAAEL"
                                              iconAttrs={
                                                Object {
                                                  "fill": "currentColor",
                                                  "xmlns": "http://www.w3.org/2000/svg",
                                                }
                                              }
                                              iconVerticalAlign="middle"
                                              iconViewBox="0 0 24 24"
                                            >
                                              <svg
                                                aria-hidden="true"
                                                className="StyledIconBase-ea9ulj-0 bsimgx sc-fFeiMQ hZAAEL"
                                                fill="currentColor"
                                                focusable="false"
                                                viewBox="0 0 24 24"
                                                xmlns="http://www.w3.org/2000/svg"
                                              >
                                                <path
                                                  d="M0 0h24v24H0z"
                                                  fill="none"
                                                />
                                                <path
                                                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2a7.2 7.2 0 01-6-3.22c.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08a7.2 7.2 0 01-6 3.22z"
                                                />
                                              </svg>
                                            </ForwardRef>
                                          </StyledIconBase>
                                        </AccountCircle>
                                      </Styled(AccountCircle)>
                                    </div>
                                  </styled.div>
                                </button>
                              </styled.button>
                            </div>
                          </styled.div>
                        </MenuButton>
                      </div>
                    </styled.div>
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </div>
        </styled.div>
      </header>
    </styled.header>
  </Header>
</Provider>
`;
